src/shared/bootspec.c

```diff
@@ -841,6 +842,107 @@ static int find_uki_sections(
         return 0;
 }
 
+static int find_addon_sections(
+                int fd,
+                const char *path,
+                char **ret_cmdline) {
+
+        _cleanup_free_ IMAGE_SECTION_HEADER *sections = NULL;
+        _cleanup_free_ PeHeader *pe_header = NULL;
+        int r;
+
+        r = find_sections(fd, path, &sections, &pe_header);
+        if (r < 0)
+                return r;
+
+        r = find_cmdline_section(fd, path, sections, pe_header, ret_cmdline);
+        if (r < 0)
+                return r;
+
+        return 0;
+}
+
+static int boot_entries_find_unified_addons(
+                BootConfig *config,
+                const char *root,
+                const char *addon_dir,
+                BootEntryAddon **addons) {
+
+        _cleanup_closedir_ DIR *d = NULL;
+        _cleanup_free_ char *full = NULL;
+        int r;
+
+        assert(addons);
+
+        r = chase_and_opendir(addon_dir, root, CHASE_PREFIX_ROOT|CHASE_PROHIBIT_SYMLINKS, &full, &d);
+        if (r == -ENOENT)
+                return 0;
+        if (r < 0)
+                return log_error_errno(r, "Failed to open '%s/%s': %m", root, addon_dir);
+
+        LIST_HEAD_INIT(*addons);
+        FOREACH_DIRENT(de, d, return log_error_errno(errno, "Failed to read %s: %m", full)) {
+                _cleanup_free_ char *j = NULL, *cmdline = NULL;
+                _cleanup_close_ int fd = -EBADF;
+                BootEntryAddon *addon = NULL;
```

Comment:
please define a proper destructor for the structure. i.e. `boot_entry_addon_free()`, and then use `DEFINE_TRIVIAL_CLEANUP_FUNC()` to generate a function that is then suitable for `_cleanup_(boot_entry_addon_freep) BootEntryAddon *addon = NULL`